spec/core/output_sink/base_sink_spec.rb

```diff
@@ -2,12 +2,162 @@
 
 describe Core::OutputSink::BaseSink do
   subject { described_class.new }
+  let(:document) { { :id => 15 } }
+  let(:deleted_id) { 25 }
 
   it 'base_sink methods raise an error when called' do
-    expect { subject.ingest(nil) }.to raise_error('not implemented')
-    expect { subject.ingest_multiple(nil) }.to raise_error('not implemented')
-    expect { subject.delete(nil) }.to raise_error('not implemented')
-    expect { subject.delete_multiple(nil) }.to raise_error('not implemented')
-    expect { subject.flush(_size: nil) }.to raise_error('not implemented')
+    expect { subject.ingest(document) }.to raise_error(NotImplementedError)
+    expect { subject.ingest_multiple([document]) }.to raise_error(NotImplementedError)
+    expect { subject.delete(deleted_id) }.to raise_error(NotImplementedError)
+    expect { subject.delete_multiple([deleted_id]) }.to raise_error(NotImplementedError)
+    expect { subject.flush }.to raise_error(NotImplementedError)
+  end
+
+  context 'when all private methods are implemented' do
+    before(:each) do
+      allow(subject).to receive(:do_ingest)
+      allow(subject).to receive(:do_delete)
+      allow(subject).to receive(:do_flush)
+      allow(subject).to receive(:do_serialize)
+    end
+
+    context '#ingest' do
+      context 'when ingested document is empty' do
+        let(:document) { {} }
+
+        it 'does not call do_ingest' do
+          expect(subject).to_not receive(:do_ingest)
+
+          subject.ingest(document)
+        end
+      end
+
+      context 'when ingested document is not empty' do
+        let(:document_id) { 15 }
+        let(:document) { { 'id' => document_id, 'something' => 'something' } }
+        let(:serialized_document) { 'id: 15, something: something' }
+
+        before(:each) do
+          allow(subject).to receive(:do_serialize).with(document).and_return(serialized_document)
+        end
+
+        it 'calls do_ingest on serialized document' do
+          expect(subject).to receive(:do_ingest).with(document_id, serialized_document)
+
+          subject.ingest(document)
+        end
+      end
+    end
+
+    context '#ingest_multiple' do
```

Comment:
[nitpick] It's better to use #describe when you start testing a method.

refs: https://www.betterspecs.org/